### PR TITLE
Events

### PR DESCRIPTION
**Pull request by warfaj**
This PR does the following:
1) Parses an event block which is a map of event keys (eg. "Click") and their associated functions. These functions can be specified as singletons or as a list. 
2) Adds the event block to the manifest
3) codegens the handler registry for "Click", "Scroll" and "WillRender" events. 
4) edits pax-example to test these new handlers

### CODE REVIEW

**Comment by warfaj** [comment]
Added in-line handlers for click and scroll.
Merged with latest master

File: pax-example/src/lib.rs [hunk 1]
```diff
@@ -1,26 +1,51 @@
...
                 Transform2D::align(100%, 0%) * Transform2D::anchor(100%, 0%)
             } />
+            <Text text="Hello world" />
+            <Path />
             <Rectangle fill={Color::rgb(0.25,0.5,0.5)} width=100% height=100% />
-       </Group>
+
+
+    @events {
+            Click: [self.handle_global_click],
```
[the remark this answers is not in the document]
Comment (zackbrown) [reply pax-example/src/lib.rs:22]: (I recall voicing a previous opinion that this should be cased as an enum member, `Click`, but seeing it changed my mind.)

**Review by zackbrown — COMMENTED** [review]
Added a commit, pushed to `origin/events`, punching through to some e2e event => property change => expression update loop. (Borrowed some patches from my current dev branch to enable this, will deal with merge later.)

1. super exciting to see scrolling hooked up with this "spinning ellipse" (pull it down & try it out!)
2. I notice that the hit box seems to be off when playing with that ellipse.  When the ellipse is at ~45deg, it doesn't respond to events at the far edges.  Do you think this is related to these latest ray cast logic changes?  Could definitely be an outstanding existing bug, though I recall kicking the tires around `transform & bounds` ca. Jabberwocky.

File: pax-example/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (warfaj) [reply pax-example/src/lib.rs:22]: yeah I agree! We should reconcile it. I'm good with eiter. I think lowercase feels better.

File: pax-core/src/runtime.rs [hunk 2]
```diff
@@ -168,24 +168,24 @@ impl<R: 'static + RenderContext> StackFrame<R> {
         }
     }
 
-    // Traverses stack recursively `n` times to retrieve descendant;
+    // Traverses stack recursively `n` times to retrieve ancestor;
     // useful for runtime lookups for identifiers
-    pub fn nth_descendant(&self, n: isize) -> Option<Rc<RefCell<StackFrame<R>>>> {
+    pub fn nth_ancestor(&self, n: isize) -> Option<Rc<RefCell<StackFrame<R>>>> {
```
[the remark this answers is not in the document]
Comment (warfaj) [reply pax-core/src/runtime.rs:173]: great thanks!

**Comment by warfaj** [comment]
With regards to the ellipse hitbox, could it be the hit test is inaccurate for ellipses?  (i.e. assumes rect-linear)

**Comment by zackbrown** [comment]
> With regards to the ellipse hitbox, could it be the hit test is inaccurate for ellipses? (i.e. assumes rect-linear)

Negative, that would effectively _increase_ the hit bounds, but in this case the bounds are drastically reduced at certain angles. It's a familiar kind of brokenness: seems to be broken transform math.  Likely there was a regression or latent bug in the `transform and bounds` a.k.a. `tab` calculation used by the hit-testing logic (could result from changes in render tree recursion order), or could be related to the z-index-sorting logic.

**Comment by zackbrown** [comment]
Dug in a bit more; I don't see anything obvious that might cause that regression re: hit box, so let's treat it as a separate bug — merging!

**Review by zackbrown — APPROVED** [review]
LGTM

**Pull request merged**